src/main/java/com/google/ehub/data/ProfileDatastore.java

```diff
@@ -83,6 +84,29 @@ public UserProfile getUserProfile(String email) {
     return createUserProfileFromEntity(userEntity);
   }
 
+  /**
+   * Finds a user profile based on unique username.
+   *
+   * @param username the value of the username property to search for
+   * @return the UserProfile found in Datastore wrapped in an {@link
+   *     Optional}, the
+   * optional object will be empty if the user Entity was not found
+   */
+  public Optional<UserProfile> queryProfileByUsername(String username) {
+    Query query =
+        new Query(PROFILE_ITEM_KIND)
+            .setFilter(new FilterPredicate(USERNAME_PROPERTY_KEY, FilterOperator.EQUAL, username));
+    PreparedQuery queryResults = datastore.prepare(query);
+
+    Entity userEntity = queryResults.asSingleEntity();
+
+    if (userEntity == null) {
+      return Optional.empty();
+    }
+
+    return Optional.of(createUserProfileFromEntity(userEntity));
```

Comment:
You can simplify this a bit using `Optional.ofNullable()`

```suggestion
    return Optional.ofNullable(createUserProfileFromEntity(userEntity));
```